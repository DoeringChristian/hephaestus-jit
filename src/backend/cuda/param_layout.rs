use std::collections::HashMap;

use crate::trace::{Trace, VarId, VarType};

#[derive(Clone, Debug, Default)]
pub struct ParamLayout(HashMap<VarId, usize>);

impl ParamLayout {
    pub fn byte_offset(&self, id: VarId) -> usize {
        self.0[&id] * std::mem::size_of::<u64>()
    }
    pub fn buffer_size(&self) -> usize {
        2 * 8
    }
    pub fn generate(trace: &Trace) -> Self {
        let mut offset: usize = 0;

        let offsets = trace
            .arrays
            .iter()
            .map(|id| {
                let o = offset;
                offset += 1;
                (*id, o)
            })
            .collect::<HashMap<_, _>>();

        Self(offsets)
    }
}
