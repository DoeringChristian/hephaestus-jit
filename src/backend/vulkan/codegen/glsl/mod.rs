use half::f16;

use super::CompileInfo;
use crate::ir::{VarId, IR};
use crate::vartype::{AsVarType, VarType};
use std::collections::{HashMap, HashSet};
use std::fmt::Write;

pub fn assemble_ir(ir: &IR, info: &CompileInfo, entry_point: &str) -> Option<Vec<u32>> {
    let mut s = String::new();
    assemble_entry_point(&mut s, ir, info, entry_point).unwrap();

    // log::trace!("\n{s}");
    print!("{s}");

    let mut options = shaderc::CompileOptions::new().unwrap();
    options.set_optimization_level(shaderc::OptimizationLevel::Performance);
    let mut compiler = shaderc::Compiler::new().unwrap();

    let artefact = compiler
        .compile_into_spirv(
            &s,
            shaderc::ShaderKind::Compute,
            "",
            entry_point,
            Some(&options),
        )
        .unwrap();

    Some(artefact.as_binary().to_vec())
}

pub fn assemble_entry_point(
    s: &mut String,
    ir: &IR,
    info: &CompileInfo,
    entry_point: &str,
) -> std::fmt::Result {
    write!(
        s,
        r#"
#version 450

// Explicit arythmetic types
#extension GL_EXT_shader_explicit_arithmetic_types: require
#extension GL_EXT_shader_explicit_arithmetic_types_int8: require
#extension GL_EXT_shader_explicit_arithmetic_types_int16: require
#extension GL_EXT_shader_explicit_arithmetic_types_int32: require
#extension GL_EXT_shader_explicit_arithmetic_types_int64: require
#extension GL_EXT_shader_explicit_arithmetic_types_float16: require
#extension GL_EXT_shader_explicit_arithmetic_types_float32: require
#extension GL_EXT_shader_explicit_arithmetic_types_float64: require

// Atomics and Memory
#extension GL_KHR_memory_scope_semantics: require
#extension GL_EXT_shader_atomic_int64: require
        
"#
    )?;

    let mut buffer_types = HashSet::new();
    bindings(s, ir, &mut buffer_types)?;
    let ty = u32::var_ty();
    if !buffer_types.contains(ty) {
        let n_buffers = ir.n_buffers + 1;
        writeln!(
            s,
            "layout(set = 0, binding = 0) buffer Buffer_{name}{{ {name} b[]; }} buffer_{name}[{n_buffers}];",
            name = GlslTypeName(ty)
        )?;
    }

    writeln!(s, "")?;
    writeln!(
        s,
        "layout(local_size_x = {workgroup_size}, local_size_y = 1, local_size_z = 1)in;",
        workgroup_size = info.work_group_size
    )?;
    writeln!(s, "void {entry_point}(){{")?;

    // Early return
    writeln!(s, "\tuint size = buffer_uint32_t[0].b[0];")?;
    writeln!(s, "\tuint index = uint(gl_GlobalInvocationID.x);")?;
    writeln!(s, "\tif (index >= size) {{return;}}")?;

    assemble_vars(s, ir)?;

    writeln!(s, "}}")?;

    Ok(())
}

fn bindings(
    s: &mut String,
    ir: &IR,
    buffer_types: &mut HashSet<&'static VarType>,
) -> std::fmt::Result {
    let n_buffers = ir.n_buffers + 1;
    for id in ir.var_ids() {
        let var = ir.var(id);
        match var.op {
            crate::op::KernelOp::BufferRef => {
                if !buffer_types.contains(var.ty) {
                    writeln!(
                        s,
                        "layout(set = 0, binding = 0) buffer Buffer_{name}{{ {name} b[]; }} buffer_{name}[{n_buffers}];",
                        name = GlslTypeName(var.ty)
                    )?;
                    buffer_types.insert(var.ty);
                }
            }
            crate::op::KernelOp::TextureRef { dim } => todo!(),
            crate::op::KernelOp::AccelRef => todo!(),
            _ => {}
        }
    }
    Ok(())
}
pub struct GlslTypeName(&'static VarType);
impl std::fmt::Display for GlslTypeName {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self.0 {
            VarType::Void => write!(f, "void"),
            VarType::Bool => write!(f, "bool"),
            VarType::I8 => write!(f, "int8_t"),
            VarType::U8 => write!(f, "uint8_t"),
            VarType::I16 => write!(f, "int16_t"),
            VarType::U16 => write!(f, "uint16_t"),
            VarType::I32 => write!(f, "int32_t"),
            VarType::U32 => write!(f, "uint32_t"),
            VarType::I64 => write!(f, "int64_t"),
            VarType::U64 => write!(f, "uint64_t"),
            VarType::F16 => write!(f, "float16_t"),
            VarType::F32 => write!(f, "float32_t"),
            VarType::F64 => write!(f, "float64_t"),
            VarType::Vec { ty, num } => match ty {
                VarType::I8 => write!(f, "i8vec{num}"),
                VarType::U8 => write!(f, "u8vec{num}"),
                VarType::I16 => write!(f, "i16{num}"),
                VarType::U16 => write!(f, "u16vec{num}"),
                VarType::I32 => write!(f, "i32vec{num}"),
                VarType::U32 => write!(f, "u32vec{num}"),
                VarType::I64 => write!(f, "i64vec{num}"),
                VarType::U64 => write!(f, "u64vec{num}"),
                VarType::F16 => write!(f, "f16vec{num}"),
                VarType::F32 => write!(f, "f32vec{num}"),
                VarType::F64 => write!(f, "f64vec{num}"),
                _ => todo!(),
            },
            VarType::Array { ty, num } => todo!(),
            VarType::Mat { ty, rows, cols } => todo!(),
            VarType::Struct { tys } => todo!(),
        }
    }
}

pub struct Reg(VarId);
impl std::fmt::Display for Reg {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "r{id}", id = self.0 .0)
    }
}

fn assemble_vars(s: &mut String, ir: &IR) -> std::fmt::Result {
    for id in ir.var_ids() {
        let var = ir.var(id);
        let reg = id.0;
        let deps = ir.deps(id);

        let ty = var.ty;
        let glsl_ty = GlslTypeName(ty);
        // let ty = GlslTypeName(var.ty);

        match var.op {
            crate::op::KernelOp::Nop => {
                writeln!(
                    s,
                    "\t{glsl_ty} {dst} = {src};",
                    dst = Reg(id),
                    src = Reg(deps[0])
                )?;
            }
            crate::op::KernelOp::Scatter(_) => {
                let dst = deps[0];
                let src = deps[1];
                let idx = deps[2];
                let glsl_ty = GlslTypeName(ir.var(src).ty);
                let buffer_idx = ir.var(dst).data + 1;

                writeln!(
                    s,
                    "\tbuffer_{glsl_ty}[{buffer_idx}].b[{idx}] = {src};",
                    idx = Reg(idx),
                    src = Reg(src)
                )?;
            }
            crate::op::KernelOp::Gather => {
                let src = deps[0];
                let idx = deps[1];
                writeln!(
                    s,
                    "\t{glsl_ty} {dst} = buffer_{glsl_ty}[{buffer_idx}].b[{idx}];",
                    dst = Reg(id),
                    buffer_idx = ir.var(src).data + 1,
                    idx = Reg(deps[1]),
                )?;
            }
            crate::op::KernelOp::Index => {
                // writeln!(
                //     s,
                //     "\tuint32_t {dst} = uint32_t(gl_GlobalInvocationID.x);",
                //     dst = Reg(id)
                // )?;
                writeln!(s, "\tuint32_t {dst} = index;", dst = Reg(id))?;
            }
            crate::op::KernelOp::Literal => {
                let dst = Reg(id);
                match ty {
                    VarType::Bool => {
                        writeln!(
                            s,
                            "\t{glsl_ty} {dst} = {glsl_ty}({data});",
                            data = if var.data == 0 { false } else { true }
                        )?;
                    }
                    VarType::I8
                    | VarType::U8
                    | VarType::I16
                    | VarType::U16
                    | VarType::I32
                    | VarType::U32
                    | VarType::I64
                    | VarType::U64 => {
                        let data = var.data;
                        writeln!(s, "\t{glsl_ty} {dst} = {glsl_ty}({data}ul);",)?;
                    }
                    VarType::F16 => {
                        let data: f16 = unsafe { *(&var.data as *const _ as *const _) };
                        writeln!(s, "\t{glsl_ty} {dst} = {glsl_ty}({data});",)?;
                    }
                    VarType::F32 => {
                        let data: f32 = unsafe { *(&var.data as *const _ as *const _) };
                        writeln!(s, "\t{glsl_ty} {dst} = {glsl_ty}({data});",)?;
                    }
                    VarType::F64 => {
                        let data: f64 = unsafe { *(&var.data as *const _ as *const _) };
                        writeln!(s, "\t{glsl_ty} {dst} = {glsl_ty}({data});",)?;
                    }
                    _ => todo!(),
                    // VarType::Vec { ty, num } => todo!(),
                    // VarType::Array { ty, num } => todo!(),
                    // VarType::Mat { ty, rows, cols } => todo!(),
                    // VarType::Struct { tys } => todo!(),
                }
            }
            crate::op::KernelOp::Extract(_) => todo!(),
            crate::op::KernelOp::Construct => todo!(),
            crate::op::KernelOp::Select => todo!(),
            crate::op::KernelOp::TexLookup => todo!(),
            crate::op::KernelOp::TraceRay => todo!(),
            crate::op::KernelOp::Bop(op) => {
                let dst = Reg(id);
                let lhs = Reg(deps[0]);
                let rhs = Reg(deps[1]);
                match op {
                    crate::op::Bop::Add => {
                        writeln!(s, "\t{glsl_ty} {dst} = {lhs} + {rhs};")?;
                    }
                    crate::op::Bop::Sub => todo!(),
                    crate::op::Bop::Mul => todo!(),
                    crate::op::Bop::Div => todo!(),
                    crate::op::Bop::Modulus => todo!(),
                    crate::op::Bop::Min => todo!(),
                    crate::op::Bop::Max => todo!(),
                    crate::op::Bop::And => todo!(),
                    crate::op::Bop::Or => todo!(),
                    crate::op::Bop::Xor => todo!(),
                    crate::op::Bop::Shl => todo!(),
                    crate::op::Bop::Shr => todo!(),
                    crate::op::Bop::Eq => todo!(),
                    crate::op::Bop::Neq => todo!(),
                    crate::op::Bop::Lt => todo!(),
                    crate::op::Bop::Le => todo!(),
                    crate::op::Bop::Gt => todo!(),
                    crate::op::Bop::Ge => todo!(),
                }
            }
            crate::op::KernelOp::Uop(_) => todo!(),
            _ => {}
        }
    }
    Ok(())
}
