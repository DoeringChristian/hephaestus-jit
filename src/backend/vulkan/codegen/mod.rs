use crate::ir::IR;

mod glsl;
mod rspirv;

#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
pub struct CompileInfo {
    pub work_group_size: u32,
}

pub fn assemble_trace(ir: &IR, info: &CompileInfo, entry_point: &str) -> Vec<u32> {
    rspirv::assemble_trace(ir, info, entry_point).unwrap()
    // glsl::assemble_ir(ir, info, entry_point).unwrap()
}
