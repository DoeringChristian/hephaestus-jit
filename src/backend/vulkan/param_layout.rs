use std::collections::HashMap;

use crate::trace::{Trace, VarId, VarType};

#[derive(Clone, Debug, Default)]
pub struct ParamLayout(HashMap<VarId, usize>, usize);

impl ParamLayout {
    pub fn buffer_idx(&self, id: VarId) -> usize {
        self.0[&id]
    }
    pub fn len(&self) -> usize {
        self.1
    }
    pub fn generate(trace: &Trace) -> Self {
        let mut i = 0;

        const BININGS_PER_SET: usize = 6;

        let buffer_idxs = trace
            .arrays
            .iter()
            .map(|id| {
                let buffer_idx = i;
                // let set = i / BININGS_PER_SET;
                // let binding = i % BININGS_PER_SET;
                i += 1;
                (*id, buffer_idx)
            })
            .collect::<HashMap<_, _>>();

        Self(buffer_idxs, i)
    }
}
