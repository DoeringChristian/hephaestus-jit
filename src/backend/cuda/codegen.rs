use crate::ir::*;
use crate::op::Op;
use crate::vartype::VarType;

/// Offset for the special registers:
///
/// Special registers:
///   %r0   :  Index
///   %r1   :  Step
///   %r2   :  Size
///   %p0   :  Stopping predicate
///   %rd0  :  Temporary for parameter pointers
///   %rd1  :  Pointer to parameter table in global memory if too big
///   %b3, %w3, %r3, %rd3, %f3, %d3, %p3: reserved for use in compound
///   statements that must write a temporary result to a register.
pub const REGISTER_OFFSET: usize = 4;

use super::param_layout::ParamLayout;
pub fn prefix(ty: &VarType) -> &'static str {
    match ty {
        VarType::Void => "%u",
        VarType::Bool => "%p",
        VarType::I8 => "%b",
        VarType::U8 => "%b",
        VarType::I16 => "%w",
        VarType::U16 => "%w",
        VarType::I32 => "%r",
        VarType::U32 => "%r",
        VarType::I64 => "%rd",
        VarType::U64 => "%rd",
        // VarType::F16 => "%h",
        VarType::F32 => "%f",
        VarType::F64 => "%d",
        _ => todo!(),
    }
}
// Retuns the cuda/ptx Representation for this type
pub fn tyname(ty: &VarType) -> &'static str {
    match ty {
        VarType::Void => "???",
        VarType::Bool => "pred",
        VarType::I8 => "s8",
        VarType::U8 => "u8",
        VarType::I16 => "s16",
        VarType::U16 => "u16",
        VarType::I32 => "s32",
        VarType::U32 => "u32",
        VarType::I64 => "s64",
        VarType::U64 => "u64",
        // VarType::F16 => "f16",
        VarType::F32 => "f32",
        VarType::F64 => "f64",
        _ => todo!(),
    }
}
pub fn tyname_bin(ty: &VarType) -> &'static str {
    match ty {
        VarType::Void => "???",
        VarType::Bool => "pred",
        VarType::I8 => "b8",
        VarType::U8 => "b8",
        VarType::I16 => "b16",
        VarType::U16 => "b16",
        VarType::I32 => "b32",
        VarType::U32 => "b32",
        VarType::I64 => "b64",
        VarType::U64 => "b64",
        // VarType::F16 => "b16",
        VarType::F32 => "b32",
        VarType::F64 => "b64",
        _ => todo!(),
    }
}

#[derive(Debug, Clone, Copy)]
pub struct Reg<'a> {
    reg: usize,
    ty: &'a VarType,
}

impl<'a> Reg<'a> {
    /// Create a register constructor, which can be called to generate the `Reg` Representation,
    /// without having to use the trace.
    ///
    /// * `trace`: Trace to generate code for
    pub fn constructor(trace: &'a IR) -> impl Fn(VarId) -> Reg<'a> {
        move |id: VarId| Reg {
            reg: id.0 + REGISTER_OFFSET,
            ty: &trace.var_ty(id),
        }
    }
}

impl<'a> std::fmt::Display for Reg<'a> {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        write!(f, "{}{}", prefix(&self.ty), self.reg)
    }
}

pub fn assemble_trace(
    asm: &mut impl std::fmt::Write,
    trace: &IR,
    entry_point: &str,
    param_ty: &str,
    // param_layout: &ParamLayout,
) -> std::fmt::Result {
    let param_layout = ParamLayout::generate(trace);
    let n_regs = trace.vars.len() + REGISTER_OFFSET; // Add 4 utility registers

    // Generate Code
    writeln!(asm, ".version {}.{}", 8, 0)?;
    writeln!(asm, ".target {}", "sm_86")?;
    writeln!(asm, ".address_size 64")?;

    writeln!(asm, "")?;

    writeln!(asm, ".entry {}(", entry_point)?;
    writeln!(
        asm,
        "\t.param .align 8 .b8 params[{param_size}]) {{",
        param_size = param_layout.buffer_size(),
    )?;
    writeln!(asm, "")?;

    writeln!(
        asm,
        "\t.reg.b8   %b <{n_regs}>; .reg.b16 %w<{n_regs}>; .reg.b32 %r<{n_regs}>;"
    )?;
    writeln!(
        asm,
        "\t.reg.b64  %rd<{n_regs}>; .reg.f32 %f<{n_regs}>; .reg.f64 %d<{n_regs}>;"
    )?;
    writeln!(asm, "\t.reg.pred %p <{n_regs}>;")?;
    writeln!(asm, "")?;

    write!(
        asm,
        "\tmov.u32 %r0, %ctaid.x;\n\
            \tmov.u32 %r1, %ntid.x;\n\
            \tmov.u32 %r2, %tid.x;\n\
            \tmad.lo.u32 %r0, %r0, %r1, %r2; // r0 <- Index\n"
    )?;

    writeln!(asm, "")?;

    writeln!(
        asm,
        "\t// Index Conditional (jump to done if Index >= Size)."
    )?;
    writeln!(
        asm,
        "\tld.param.u32 %r2, [params]; // r2 <- params[0] (Size)"
    )?;
    writeln!(
        asm,
        "\tld.param.u64 %rd1, [params+8]; // rd1 <- params[8] (parameter table)"
    )?;

    write!(
        asm,
        "\tsetp.ge.u32 %p0, %r0, %r2; // p0 <- r0 >= r2\n\
        \t@%p0 bra done; // if p0 => done\n\
        \t\n\
        \tmov.u32 %r3, %nctaid.x; // r3 <- nctaid.x\n\
        \tmul.lo.u32 %r1, %r3, %r1; // r1 <- r3 * r1\n\
        \t\n"
    )?;

    write!(asm, "body: // sm_{}\n", 86)?; // TODO: compute capability from device

    for var in trace.var_ids() {
        assemble_var(asm, trace, var, param_ty, &param_layout)?;
    }

    // End of kernel:

    writeln!(asm, "\n\t//End of Kernel:")?;
    writeln!(
        asm,
        "\n\tadd.u32 %r0, %r0, %r1; // r0 <- r0 + r1\n\
           \tsetp.ge.u32 %p0, %r0, %r2; // p0 <- r1 >= r2\n\
           \t@!%p0 bra body; // if p0 => body\n\
           \n"
    )?;
    writeln!(asm, "done:")?;
    write!(
        asm,
        "\n\tret;\n\
       }}\n"
    )?;
    Ok(())
}

pub fn assemble_var(
    asm: &mut impl std::fmt::Write,
    trace: &IR,
    varid: VarId,
    param_ty: &str,
    param_layout: &ParamLayout,
) -> std::fmt::Result {
    let reg = Reg::constructor(trace);
    let deps = trace.deps(varid);

    // Write out debug info:
    writeln!(asm, "")?;

    let op = &trace.var(varid).op;
    writeln!(asm, "// {:?}:", op)?;
    match &op {
        Op::Nop => {}
        Op::Bop(bop) => {
            let lhs = deps[0];
            let rhs = deps[1];
            match bop {
                crate::op::Bop::Add => {
                    writeln!(
                        asm,
                        "\tadd.{} {}, {}, {};",
                        tyname(trace.var_ty(varid)),
                        reg(varid),
                        reg(lhs),
                        reg(rhs)
                    )?;
                }
            }
        }
        Op::Scatter => {
            let dst = deps[0];
            let src = deps[1];
            let idx = deps[2];

            let dst_var = trace.var(dst);

            // assert_eq!(dst_var.ty, VarType::Array);

            let param_offset = param_layout.byte_offset(dst);

            writeln!(
                asm,
                "\tld.global.u64 %rd0, [%rd1+{param_offset}]; //Load buffer pointer from table"
            )?;

            // Multiply idx with type size and add ptr
            let ty = trace.var_ty(src);
            writeln!(
                asm,
                "\tmad.wide.{ty} %rd3, {idx}, {ty_size}, %rd0;",
                ty = tyname(trace.var_ty(idx)),
                idx = reg(idx),
                ty_size = ty.size(),
            )?;

            let op_type = "st";
            let op = "";
            writeln!(
                asm,
                "\t{}.global{}.{} [%rd3], {};",
                op_type,
                op,
                tyname(ty),
                reg(src),
            )?;
        }
        Op::Gather => {
            let src = deps[0];
            let idx = deps[1];

            let src_var = trace.var(src);

            // Load array ptr:

            let param_offset = param_layout.byte_offset(src);

            writeln!(
                asm,
                "\tld.global.u64 %rd0, [%rd1+{param_offset}]; //Load buffer pointer from table"
            )?;

            // Multiply idx with type size and add ptr
            let ty = trace.var_ty(varid);
            writeln!(
                asm,
                "\tmad.wide.{ty} %rd3, {idx}, {ty_size}, %rd0;",
                ty = tyname(trace.var_ty(idx)),
                idx = reg(idx),
                ty_size = ty.size(),
            )?;

            // Load value from buffer
            writeln!(
                asm,
                "\tld.global.nc.{ty} {ptr}, [%rd3];",
                ty = tyname(ty),
                ptr = reg(varid),
            )?;
        }
        Op::Index => {
            let ty = trace.var_ty(varid);
            writeln!(asm, "\tmov.{} {}, %r0;", tyname(ty), reg(varid))?;
        }
        Op::Literal => {
            // let ty = trace.var_ty(varid);
            //
            // writeln!(
            //     asm,
            //     "\tmov.{tyname} {dst}, 0x{data:x};\n",
            //     tyname = tyname_bin(trace.var_ty(varid)),
            //     dst = reg(varid),
            //     data = data,
            // )?;
            todo!()
        }
        Op::Buffer => {}
        _ => {}
    };
    Ok(())
}
