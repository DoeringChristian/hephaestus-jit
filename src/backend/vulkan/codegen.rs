use std::collections::HashMap;
use std::ops::{Deref, DerefMut};

use crate::ir::{VarId, IR};
use crate::op::Op;
use crate::vartype::VarType;
use rspirv::binary::{Assemble, Disassemble};
use rspirv::{dr, spirv};

use super::param_layout::ParamLayout;

// fn ty(ty: &VarType) ->

pub fn assemble_trace(trace: &IR, entry_point: &str) -> Result<Vec<u32>, dr::Error> {
    let mut b = SpirvBuilder::default();
    b.assemble(trace, entry_point)?;

    let module = b.module();
    print!("{}", module.disassemble());
    Ok(module.assemble())
}
fn isfloat(ty: &VarType) -> bool {
    match ty {
        VarType::F32 | VarType::F64 => true,
        _ => false,
    }
}
fn isint(ty: &VarType) -> bool {
    match ty {
        VarType::I8
        | VarType::U8
        | VarType::I16
        | VarType::U16
        | VarType::I32
        | VarType::U32
        | VarType::I64
        | VarType::U64 => true,
        _ => false,
    }
}

#[derive(Default)]
struct SpirvBuilder {
    b: dr::Builder,
    spirv_vars: Vec<u32>,
}
impl Deref for SpirvBuilder {
    type Target = dr::Builder;

    fn deref(&self) -> &Self::Target {
        &self.b
    }
}
impl DerefMut for SpirvBuilder {
    fn deref_mut(&mut self) -> &mut Self::Target {
        &mut self.b
    }
}

impl SpirvBuilder {
    pub fn module(self) -> dr::Module {
        self.b.module()
    }
    pub fn assemble(&mut self, trace: &IR, entry_point: &str) -> Result<(), dr::Error> {
        // let param_layout = ParamLayout::generate(trace);
        // dbg!(&param_layout);

        self.id();

        self.acquire_ids(trace);

        self.set_version(1, 5);
        self.capability(spirv::Capability::Shader);
        self.memory_model(spirv::AddressingModel::Logical, spirv::MemoryModel::GLSL450);

        let void = self.type_void();
        let voidf = self.type_function(void, vec![]);

        let u32_ty = self.type_int(32, 0);
        let v3u32_ty = self.type_vector(u32_ty, 3);
        let ptr_v3u32_ty = self.type_pointer(None, spirv::StorageClass::Input, v3u32_ty);
        let global_invocation_id =
            self.variable(ptr_v3u32_ty, None, spirv::StorageClass::Input, None);

        self.decorate(
            global_invocation_id,
            spirv::Decoration::BuiltIn,
            [dr::Operand::BuiltIn(spirv::BuiltIn::GlobalInvocationId)],
        );

        let storage_vars = self.assemble_storage_vars(trace);

        let func = self.begin_function(
            void,
            None,
            spirv::FunctionControl::DONT_INLINE | spirv::FunctionControl::CONST,
            voidf,
        )?;

        self.begin_block(None)?;

        self.assemble_vars(trace, global_invocation_id)?;

        self.ret()?;
        self.end_function()?;

        let interface = [global_invocation_id]
            .into_iter()
            .chain(storage_vars.into_iter())
            .collect::<Vec<_>>();

        self.entry_point(
            spirv::ExecutionModel::GLCompute,
            func,
            entry_point,
            interface,
        );
        //OpEntryPoint GLCompute %main "main" %test %gl_GlobalInvocationID %test2
        self.execution_mode(func, spirv::ExecutionMode::LocalSize, [1, 1, 1]);

        Ok(())
    }
    fn acquire_ids(&mut self, trace: &IR) {
        let vars = trace.vars.iter().map(|_| self.id()).collect::<Vec<_>>();
        self.spirv_vars = vars;
    }
    fn get(&self, id: VarId) -> u32 {
        self.spirv_vars[id.0]
    }
    fn spirv_ty(&mut self, ty: &VarType) -> u32 {
        match ty {
            VarType::Void => self.type_void(),
            VarType::Bool => self.type_bool(),
            VarType::I8 => self.type_int(8, 1),
            VarType::U8 => self.type_int(8, 0),
            VarType::I16 => self.type_int(16, 1),
            VarType::U16 => self.type_int(16, 0),
            VarType::I32 => self.type_int(32, 1),
            VarType::U32 => self.type_int(32, 0),
            VarType::I64 => self.type_int(64, 1),
            VarType::U64 => self.type_int(64, 0),
            VarType::F32 => self.type_float(32),
            VarType::F64 => self.type_float(64),
        }
    }
    // TODO: Spirv Builder
    fn assemble_storage_vars(&mut self, ir: &IR) -> Vec<u32> {
        ir.var_ids()
            .filter_map(|varid| {
                let var = ir.var(varid);
                match var.op {
                    Op::Buffer => {
                        let ty = self.spirv_ty(&var.ty);
                        let u32_ty = self.type_int(32, 0);
                        let array_len = self.constant_u32(u32_ty, ir.n_buffers as _);
                        let rta_ty = self.type_runtime_array(ty);
                        let struct_ty = self.type_struct([rta_ty]);
                        let array_ty = self.type_array(struct_ty, array_len);

                        self.decorate(struct_ty, spirv::Decoration::Block, []);
                        self.member_decorate(
                            struct_ty,
                            0,
                            spirv::Decoration::Offset,
                            [dr::Operand::LiteralInt32(0)],
                        );
                        self.decorate(
                            rta_ty,
                            rspirv::spirv::Decoration::ArrayStride,
                            [dr::Operand::LiteralInt32(4)],
                        );

                        let ptr_ty =
                            self.type_pointer(None, spirv::StorageClass::StorageBuffer, array_ty);

                        let dst = self.get(varid);

                        self.variable(ptr_ty, Some(dst), spirv::StorageClass::StorageBuffer, None);

                        self.decorate(
                            dst,
                            spirv::Decoration::DescriptorSet,
                            [dr::Operand::LiteralInt32(0)],
                        );
                        self.decorate(
                            dst,
                            spirv::Decoration::Binding,
                            [dr::Operand::LiteralInt32(0)],
                        );
                        Some(dst)
                    }
                    _ => None,
                }
            })
            .collect()
    }

    fn assemble_vars(&mut self, ir: &IR, global_invocation_id: u32) -> Result<(), dr::Error> {
        for varid in ir.var_ids() {
            let var = ir.var(varid);
            let deps = ir.deps(varid);
            match var.op {
                Op::Nop => {}
                Op::Add => {
                    let dst = self.get(varid);
                    let lhs = self.get(deps[0]);
                    let rhs = self.get(deps[1]);
                    let ty = self.spirv_ty(&var.ty);
                    if isint(&var.ty) {
                        self.i_add(ty, Some(dst), lhs, rhs)?;
                    } else if isfloat(&var.ty) {
                        self.f_add(ty, Some(dst), lhs, rhs)?;
                    } else {
                        todo!()
                    }
                }
                Op::Scatter => {
                    let dst = deps[0];
                    let src = deps[1];
                    let idx = deps[2];
                    dbg!(&ir.var(src));
                    dbg!(&ir.var(dst));

                    let buffer_idx = ir.var(dst).data;

                    let ty = self.spirv_ty(ir.var_ty(src));
                    let ptr_ty = self.type_pointer(None, spirv::StorageClass::StorageBuffer, ty);
                    let int_ty = self.type_int(32, 0);
                    let buffer = self.constant_u32(int_ty, buffer_idx as _);
                    let elem = self.constant_u32(int_ty, 0);

                    let dst = self.get(dst);
                    let idx = self.get(idx);
                    let src = self.get(src);

                    let ptr = self.access_chain(ptr_ty, None, dst, [buffer, elem, idx])?;
                    self.store(ptr, src, None, None)?;
                }
                Op::Gather => {
                    let src = deps[0];
                    let idx = deps[1];
                    let buffer_idx = ir.var(src).data;

                    let ty = self.spirv_ty(&var.ty);
                    let ptr_ty = self.type_pointer(None, spirv::StorageClass::StorageBuffer, ty);
                    let int_ty = self.type_int(32, 0);
                    let buffer = self.constant_u32(int_ty, buffer_idx as _);
                    let elem = self.constant_u32(int_ty, 0);

                    let src = self.get(src);
                    let idx = self.get(idx);
                    let dst = self.get(varid);

                    let ptr = self.access_chain(ptr_ty, None, src, [buffer, elem, idx])?;
                    self.load(ty, Some(dst), ptr, None, None)?;
                }
                Op::Index => {
                    let u32_ty = self.type_int(32, 0);
                    let ptr_ty = self.type_pointer(None, spirv::StorageClass::Input, u32_ty);
                    let u32_0 = self.constant_u32(u32_ty, 0);
                    let ptr = self.access_chain(ptr_ty, None, global_invocation_id, [u32_0])?;

                    let dst = self.get(varid);

                    self.load(u32_ty, Some(dst), ptr, None, None)?;
                }
                Op::Literal => {
                    let ty = self.spirv_ty(&var.ty);
                    let c = match &var.ty {
                        VarType::Bool => {
                            if var.data == 0 {
                                self.constant_false(ty)
                            } else {
                                self.constant_true(ty)
                            }
                        }
                        VarType::I32 | VarType::U32 => self.constant_u32(ty, var.data as _),
                        VarType::I64 | VarType::U64 => self.constant_u64(ty, var.data),
                        // VarType::F32 => self.constant_f32(ty, var.data),
                        // VarType::F64 => self.constant_f64(ty, var.data),
                        _ => todo!(),
                    };
                    self.spirv_vars[varid.0] = c;
                }
                _ => {}
            }
        }
        Ok(())
    }
}
