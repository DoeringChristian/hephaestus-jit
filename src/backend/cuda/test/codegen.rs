use crate::backend::cuda::codegen::assemble_trace;
use crate::backend::cuda::param_layout::ParamLayout;
use crate::trace::*;

#[test]
fn gather() {
    let mut trace = Trace::default();

    let array = trace.push_var(Var {
        ty: VarType::Array,
        external: Some(0),
    });

    let x = trace.push_var(Var {
        ty: VarType::U8,
        ..Default::default()
    });
    let idx = trace.push_var(Var {
        ty: VarType::U32,
        ..Default::default()
    });

    let gather = trace.push_op(Op::Gather {
        src: array,
        dst: x,
        idx,
    });

    let layout = ParamLayout {
        size: 0,
        arrays: 1,
        n_params: 2,
    };

    let mut asm = String::new();
    assemble_trace(&mut asm, &trace, gather, "global", layout).unwrap();
    insta::assert_snapshot!(asm);
}

#[test]
fn scatter() {
    let mut trace = Trace::default();

    let array = trace.push_var(Var {
        ty: VarType::Array,
        external: Some(0),
    });

    let x = trace.push_var(Var {
        ty: VarType::U8,
        ..Default::default()
    });
    let idx = trace.push_var(Var {
        ty: VarType::U32,
        ..Default::default()
    });

    let scatter = trace.push_op(Op::Scatter {
        src: x,
        dst: array,
        idx,
    });

    let layout = ParamLayout {
        size: 0,
        arrays: 1,
        n_params: 2,
    };

    let mut asm = String::new();
    assemble_trace(&mut asm, &trace, scatter, "global", layout).unwrap();
    insta::assert_snapshot!(asm);
}
