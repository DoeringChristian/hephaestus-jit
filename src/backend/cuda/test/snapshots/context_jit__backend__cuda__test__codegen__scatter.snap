---
source: src/backend/cuda/test/codegen.rs
expression: asm
---
// VarId(0)[VarId(2)] <- VarId(1):
	ld.global.u64 %rd0 [params+8]
	mad.wide.u8 %rd3, %r2, 1, %rd0;
	st.global.u8 [%rd3], %b1;

